### PR TITLE
fix(ast/estree): fix `StringLiteral`s containing lone surrogates

### PR DESCRIPTION
**Pull request by overlookmotel**
Improve output for `StringLiteral`s containing lone surrogates.

Most of the remaining failing tests are just due to difference in case e.g.:

```diff
- "value": "\ud83d",
+ "value": "\uD83D",
```

But there are still a few genuine test fails. We don't have it exactly right yet. But this is a step in the right direction.

### CODE REVIEW

File: crates/oxc_ast/src/serialize.rs [hunk 1]
```diff
@@ -225,6 +227,71 @@ impl ESTree for NullLiteralRaw<'_> {
...
+            let mut encoded = Vec::with_capacity(raw_unquoted.len());
+            let mut has_slash = false;
+            for &b in raw_unquoted.as_bytes() {
+                if b == b'\\' {
+                    has_slash = !has_slash;
+                } else if b == b'"' && !has_slash {
+                    encoded.push(b'\\');
+                }
+                encoded.push(b);
+            }
```
Comment (hi-ogawa) [inline crates/oxc_ast/src/serialize.rs:284]: Is `has_slash` correct? It looks like it's counting all previous occurrence instead of preceding sequence. 
For example, `'\uD83D"'` is edge case and it doesn't escape to `"\uD83D\""`?

(I know this is extremely rare edge cases, but I wonder if we can add extra manual test case since test262 didn't have this one).

Comment (overlookmotel) [reply crates/oxc_ast/src/serialize.rs]: You're right! I neglected to reset `has_slash = false` after a non-slash character. Fixed now.

Anyway, this is superceded by #10041.

As far as tests, it would be a pain to add an ESTree test, because it'd complicate the test runner. But we can add parser tests in #10041. There are some other new tests we need for that PR anyway.
